src/gravel/controllers/auth.py

```diff
@@ -0,0 +1,207 @@
+        return user.verify_password(password)
+
+
+class JWT(NamedTuple):
+    iss: str
+    sub: Union[str, int]
+    iat: int
+    nbf: int
+    exp: int
+    jti: str
+
+
+class JWTManager:
+    JWT_ISSUER = "Aquarium"
+    JWT_ALGORITHM = "HS256"
+    COOKIE_KEY = "access_token"
+
+    def __init__(self, config: AuthOptionsModel):
+        self._config: AuthOptionsModel = config
+
+    def create_access_token(self, subject: Union[str, int]) -> str:
+        now = int(datetime.now(timezone.utc).timestamp())
+        payload = JWT(
+            iss=self.JWT_ISSUER,
+            sub=subject,
+            iat=now,
+            nbf=now,
+            exp=now + self._config.jwt_ttl,
+            jti=str(uuid.uuid4())
+        )._asdict()
+        encoded_token = jwt.encode(
+            payload, self._config.jwt_secret,
+            algorithm=self.JWT_ALGORITHM)
+        return encoded_token
+
+    def get_raw_access_token(self, token) -> JWT:
+        raw_token = jwt.decode(token, self._config.jwt_secret,
+                               algorithms=[self.JWT_ALGORITHM])
+        return JWT(**raw_token)
+
+    @staticmethod
+    def set_token_cookie(response: Response, token: str) -> None:
+        response.set_cookie(key=JWTManager.COOKIE_KEY, value=f"Bearer {token}",
+                            httponly=True, samesite="strict")
+
+    @staticmethod
+    def get_token_from_cookie(request: Request) -> Optional[str]:
+        value: Optional[str] = request.cookies.get(JWTManager.COOKIE_KEY)
+        if value is None:
+            return None
+        scheme, token = get_authorization_scheme_param(value)
+        if scheme.lower() == "bearer":
+            return token
+        return None
+
+
+class JWTDenyList:
+    """
+    This list contains JWT tokens that are not allowed to use anymore.
+    E.g. a token is added when a user logs out of the UI. The list will
+    automatically remove expired tokens.
+    """
+
+    def __init__(self, store: KV):
+        self._store: KV = store
+        self._jti_dict: Dict[str, int] = {}
+
+    def _cleanup(self):
+        now = int(datetime.now(timezone.utc).timestamp())
+        self._jti_dict = {jti: exp for jti, exp in self._jti_dict.items() if exp > now}
+
+    async def load(self) -> None:
+        self._jti_dict = {}
+        value = await self._store.get("/auth/jwt_deny_list")
+        if value is not None:
+            self._jti_dict = json.loads(value)
+            self._cleanup()
+
+    async def save(self) -> None:
+        await self._store.put("/auth/jwt_deny_list", json.dumps(self._jti_dict))
+
+    def add(self, token: JWT) -> None:
+        self._jti_dict[token.jti] = token.exp
+
+    def includes(self, token: JWT) -> bool:
+        return token.jti in self._jti_dict
+
+
+class JWTAuthSchema(OAuth2PasswordBearer):
+    def __init__(self):
+        super().__init__(tokenUrl="auth/login")
+
+    async def __call__(self, request: Request) -> Optional[JWT]:  # type: ignore[override]
```

Comment:
The API is leaking here by passing a `Request` to the controller. It would be preferable to pass what this class actually requires (seems it's the `GlobalState`?) instead of a fastapi request.

Reply:
The method signature is given by the derived class [OAuth2PasswordBearer](https://github.com/tiangolo/fastapi/blob/master/fastapi/security/oauth2.py#L152)

Can we relocate this class to somewhere else in our source tree where we do not pollute our `controller` rules?

Reply:
If this is bound to the API, then it should probably live under `api/auth.py`.